## Final Project Submission

Please fill out:
* Student name: Brian Kanyenje
* Student pace: part time
* Scheduled project review date/time: 24/01/2025
* Instructor name: Veronica Isiaho
* Blog post URL: 


# MICROSOFT MOVIE STUDIO INSIGHT

## 1.1 Introduction

With many different companies getting into the movie industry, Microsoft has also decided to join the movement by creating its own film studio. However, asg a tech company with little to no knowledge when it comes to the film industry, Microsoft is faced with significant challenges such as: 
* identifying the types of films currently performing well
* understanding the factors that make a film successful

This project will solve these problems by analyzing the current trends in the film industry and translating those insights into tangible recommendationsi in order to make Microsoft's debut into the film industry successful. By examining what films and genres are rated higly as well as gross highly, he research will provide integral insight into the ingredients that make a film studio successful. All this tanslates to the amount of returns generated by the film studio.

The primary stakeholders of this project are the head of Microsoft's film studio and the production team. The insight generated by this project will guide the heads of the film industry to make informed decisions that will ensure the films produced align with the market demand. The insight will also be used by the production team in order to know which genres to capitalize on and which production styles to use. 

In conclusion, this project is very fundamental to Microsoft as it will provide important insight that will make their insertion into the film industry successful.

## 1.2 Objectives

The objectives of this project will be:
1. To determine which genres are performing highly in terms of revenue and rating.
2. To analyze domestic and foreign market trends.
3. To investigate the impact of runtime on film performance.
4. To identify the movie studios that consistently produce high-performing films.

## 1.3 Research Questions

The research questions for this topic will be:
1. Which genres are performing highly in terms of revenue and rating?
2. What are the market trends in the domestic and foreign film markets?
3. What is the impact of runtime on film performance?
4. Which movie studios consistently produce high-performing films?

## 1.4 Hypothesis

Alternative Hypothesis (H₀): Certain movie genres perform higher in terms of revenue and ratings compared to others.
Null Hypothesis (H₁): There is no significant difference in the film performance by different genres in terms of revenue and ratings compared to others.

## 2.1 Data Overview

**Data** is the most important part of the project since it's what will be used to provide insight and recommendations. In this project the data files that will be used are:
* imdb.title.basics
* imdb.title.ratings
* bom.movie_gross 

These files come from movie datasets:
* Box Office - This dataset provides data on how much revenue movies earn through ticket sales. This data is a direct measure of the commercial success of the films. It will provide insight for deciding the genres.
* IMDB - This database contains data about the user ratings, reviews and production details. It ise used to generate feedback about a certain film. 
* Rotten Tomatoes - This database contains information about how well a movie was received. It is used to gauge a movie's overall quality in terms of user rating and criticism
* The MovieDB - This database contains information on ratings, reviews, trailers, and images.
* The Numbers - This database provides detailed information on box office performance, film production budgets, and financial aspects of the movie industry. It tracks the financial success of films and gives insights into their profitability, including revenue from domestic and international markets.

These data sources are relevant to the project since they will provide useful insight on what makes the films successful in terms of the revenue generated by the films, their overall rating and reviews and the production details. Information such as genre popularity, user ratings and financial performance will aid Microsoft in making an infomrmed decision on te type of films to produce.

   ### a)  imdb.title.basics

Firstly, we will check what this datafile contains. This can be done by converting the data into a Pandas DataFrame.

In [95]:
# Importing pandas library for data analysis and manipulation
import pandas as pd
# Importing the csv module in order to work with csv files
import csv
# Importing the gzip module, which allows working with files compressed in gzip format
import gzip
# This will extract the file and convert it into a pandas dataframe named basics
# the compression parameter specifies that the file is compressed using gzip 
basics = pd.read_csv('zippedData/imdb.title.basics.csv.gz', compression ='gzip')

The size of the dataset can be checked using `dataframe.shape` attribute:

In [96]:
# Checking the size of the dataset
basics.shape

(146144, 6)

From this it is evident that the dataset has 146144 rows and 6 columns.


In order to get a better understanding of the dataframe and its content, the first five rows can be viewed using `.head() method`

In [155]:
# Checking the first five rows
basics.head()

tconst                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes                genres  
0        2013            175.0    Action,Crime,Drama  
1        2019            114.0       Biography,Drama  
2        2018            122.0                 Drama  
3        2018             87.0          Comedy,Drama  
4        2017             80.0  Comedy,Drama,Fantasy

This file contains the basic information on the films including their title, start year, runtime_minutes and their respective genres. 

Before performing any descriptive statistics, we need to check if there is any missing data in the dataset. This can be done using the `.isna()` method.

In [98]:
# isna() checks for any missing values and adding .sum() returns the sum of missing values in the columns
basics.isna().sum()

tconst                 0
primary_title          0
original_title        21
start_year             0
runtime_minutes    31739
genres              5408
dtype: int64

The original_title column has 21 missing values, runtime_minutes has 31739 while genres has 5408 values missing. The missing values in the original_title are very few can be replaced with the values in the primary_title without impacting the outcoming of the analysis. This can be done using the `.fillna()` method

In [99]:
# Fill missing values in the 'original_title' column with corresponding values from the 'primary_title' column
basics['original_title'] = basics['original_title'].fillna(basics['primary_title'])

One of the objectives of this project is to investigate the ideal runtime. Because of this, the missing values in runtime_minutes column cannot be dropped since this will affect the outcome of the analysis. the best approach will be to fill the missing values with the median of the column. The median is preferred and not the mean since the mean is affected by outliers.

In [100]:
# Fill missing values in the 'runrime_minutes' column with the columns' median
basics['runtime_minutes'] = basics['runtime_minutes'].fillna(basics['runtime_minutes'].median())

To deal with the missing data in the genres column, the best approach will be to drop the rows with the missing data since the number of missing values is low and will most likely not affect the outcome of the analysis. `.dropna()` method will be used.

In [104]:
# inplace=True is passed to mutate the dataframe inplace
basics.dropna(inplace=True)

Next we'll check if there are any left missing values:

In [106]:
basics.isna().sum()

tconst             0
primary_title      0
original_title     0
start_year         0
runtime_minutes    0
genres             0
dtype: int64

There aren't any left missing values in the dataset. The next step is now getting the descriptive statistics of the dataframe.

`.describe()` will be used to get the descriptive statistics of the dataframe.

In [154]:
# this generates the descriptive statistics of a dataframe
basics.describe()

start_year  runtime_minutes
count  140736.000000    140736.000000
mean     2014.613986        86.411387
std         2.735569       149.933380
min      2010.000000         1.000000
25%      2012.000000        75.000000
50%      2015.000000        87.000000
75%      2017.000000        95.000000
max      2115.000000     51420.000000

From this it can be derived that the average start year is 2014 while the average runtime is 86 minutes. The minimum start year is 2010 while the maximum start year is 2115. The maximum start year appears to be a placeholder value that will be addressed suring data cleaning. Similarly, the maximum and minimum runtimes are 51420 and 1 minute. These are also placeholder values that will be dealt with during data cleaning.

### Relevance of this data

This data is highly relevant to this project. For one, it will be used to investigate the impact of runtime and film performance. This will help Microsoft's new studio in tailoring their films's length in order to meet the viewers' preferences. It will also be used in investigating the genres that are highly successful in terms of revenue generation and rating. identifying these genres will help Microsoft in choosing which genres to focus on producing.

### b) imdb.title.ratings

The second dataset that will be used in the analysis is the imdb.title.ratings. The data is contained in a compressed file and will need to be extracted and read similarly to the **imdb.title.basics**

In [109]:
# the file wille be converted to a pandas dataframe named ratings
ratings = pd.read_csv('zippedData/imdb.title.ratings.csv.gz', compression ='gzip')

The size of the dataset can be checked:

In [110]:
ratings.shape

(73856, 3)

The dataframe has 73856 rows and 3 columns. To get a better understanding of the dataset we can check the first five rows of the dataframe:

In [111]:
ratings.head()

tconst  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

The dataframe has 3 columns namely: tconst, averagerating which represents the average rating of the movies from a scale of 1 to 10 and numvotes which represent the number of votes each movie has received. 

Next let's check the dataframe for any missing values.

In [114]:
ratings.isna().sum()

tconst           0
averagerating    0
numvotes         0
dtype: int64

Wonderful! This dataframe does not have any missing values. The next step is getting the descriptive statistics of the dataframe using the `.describe()` method:

In [112]:
ratings.describe()

averagerating      numvotes
count   73856.000000  7.385600e+04
mean        6.332729  3.523662e+03
std         1.474978  3.029402e+04
min         1.000000  5.000000e+00
25%         5.500000  1.400000e+01
50%         6.500000  4.900000e+01
75%         7.400000  2.820000e+02
max        10.000000  1.841066e+06

From this it can be derived that the mean average rating is 6.33 while the mean number of votes is 3524. The maximum average rating is 10 and the maximum number of voted is 1841066.  The minimum average rating is 1 while the minimum number of votes is 5.

### Relevance of this data
This data will be used to gauge a genre's performance based on its rating and votes. This insight will be used to guide Microsoft's new studio in choosing which type of genre to capitalize on in order to get the most ratings and generate high revenue. it will also be used in investigating which movie studios produce gihly rated films. This insight will be used by Microsoft's new studio to benchmark these studios and adopt their production styles in order to enhance their own productions.

### c) bom.movie_gross

This dataset is also stored similarly to the first two and will be extracted and read similar to the other two.

In [139]:
# the file will be converted to a pandas dataframe names gross
gross = pd.read_csv('zippedData/bom.movie_gross.csv.gz', compression ='gzip')

Let's check the size of the dataframe:

In [140]:
gross.shape

(3387, 5)

This dataframe has 3387 rows and 5 columns. These columns as well as their contents can be checked using the `.head()` method:

In [141]:
gross.head()

title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross  year  
0     652000000  2010  
1     691300000  2010  
2     664300000  2010  
3     535700000  2010  
4     513900000  2010

This dataframe has 5 columns namely: title, studio, domestic_gross, foreign_gross and year. This dataset seems to represent the revenue that each respective movie generated in the domestic and foreign film market.

Let's check the dataframe for any missing values:

In [142]:
gross.isna().sum()

title                0
studio               5
domestic_gross      28
foreign_gross     1350
year                 0
dtype: int64

There are 28 missing values in the domestic_gross column, 1350 in the foreign_gross column and 5 in the studio column. The missing values in the domestic_gross column can be replaced with the meadian:

In [143]:
gross['domestic_gross'] = gross['domestic_gross'].fillna(gross['domestic_gross'].median())

The foreign_gross column has a significant number of missing values. However, it is very important in understanding the global trends hence the rows with missing values can't be dropped. The best approach will be to replace the values with the median of the column:

In [144]:
gross['foreign_gross'] = gross['foreign_gross'].fillna(gross['foreign_gross'].median())

TypeError: 

This is producing an error "could not convert string to float: 'a'" meaning that the dataframe contains values that are strings. To solve this we'll need to change the values to numeric values.

In [145]:
# Convert the 'foreign_gross' column to numeric values. 
# Any values that cannot be converted (such as strings or invalid data) will be replaced with NaN.
gross['foreign_gross'] = pd.to_numeric(gross['foreign_gross'], errors='coerce')

We will rerun the code once more:

In [146]:
gross['foreign_gross'] = gross['foreign_gross'].fillna(gross['foreign_gross'].median())

Lastly, the rows with the missing values in the studio column can be dropped since the number of missing values is very low and won't impact the analysis.

In [147]:
gross.dropna(inplace=True)

Now let's check if there are any missing values remaining.

In [149]:
gross.isna().sum()

title             0
studio            0
domestic_gross    0
foreign_gross     0
year              0
dtype: int64

Voila! We have solved the issue of the pesky missing values. Let's now check the descriptive statistics of this dataset:

In [151]:
# Check the descriptive statistics
gross.describe()

domestic_gross  foreign_gross         year
count    3.382000e+03   3.382000e+03  3382.000000
mean     2.856106e+07   5.262386e+07  2013.959787
std      6.679161e+07   1.100744e+08     2.477735
min      1.000000e+02   6.000000e+02  2010.000000
25%      1.230000e+05   1.190000e+07  2012.000000
50%      1.400000e+06   1.890000e+07  2014.000000
75%      2.767500e+07   2.917500e+07  2016.000000
max      9.367000e+08   9.605000e+08  2018.000000

The average values for domestic_gross, foreign_gross and year are 28561060, 52623860 and 2014 respectively. The maximum domestic gross is 93670000 and the maximum foreign gross is 96050000. The minimum domestic gross is 100 and the minimum foreign gross is 200. Thiese minimum values appear to be placeholder values that will be addressed during data cleaning. 

### Relevance of this data
This data will be used to investigate the genres that are generating the most revenue. This will helpguide Microsoft's new producton team in identifying which genres to focus on in order to make thebmost revenue. The data will also be used to identify which genres are doing well in the domestic and foreign market. This will also help the head of Microsoft's new studio in identifying which genres to focus on depending on the makret targeted.

 ## 2.2 Limitations
 * The basics and gross database has the movie titles identifying the different films while the ratings dataframe has only the tconst column to identify the movies. If there is any mismatch then it may be difficult in combining these dataframes. 
 * In the basics dataframe the movies have a primary title and an original title. in some instances a movie has two different titles in this dataframe and this may complicate merging this dataframe with another. 
 * In the basics dataframe some movies have too short or too long runtime and this may impact analysis unless corrected.

## 3.1 Data Preparation

In [3]:
with gzip.open('zippedData/imdb.title.basics.csv.gz', mode='rt', encoding='utf-8') as f:
    df = csv.DictReader(f)
    title = list(df)

In [4]:
with gzip.open('zippedData/imdb.title.ratings.csv.gz', mode='rt', encoding='utf-8') as f:
    df1 = csv.DictReader(f)
    ratings = list(df1)

In [5]:
with gzip.open('zippedData/bom.movie_gross.csv.gz', mode='rt', encoding='utf-8') as f:
    df = csv.DictReader(f)
    gross = list(df)

In [6]:
titles = [i['primary_title'] for i in title]

In [11]:
gross1 = []
for j in gross:
    if j['title'] in titles:
        gross1.append(j['title'])
len(gross1)

2606

In [53]:
ratings.head().sort_values(by='tconst')

tconst  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

In [54]:
gross.head().sort_values(by='foreign_gross', ascending=False)

title studio  domestic_gross  \
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
0                                  Toy Story 3     BV     415000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross  year  
1     691300000  2010  
2     664300000  2010  
0     652000000  2010  
3     535700000  2010  
4     513900000  2010